--- a/drivers/net/ethernet/ralink/esw_rt3052.c
+++ b/drivers/net/ethernet/ralink/esw_rt3052.c
@@ -721,10 +721,14 @@ static void esw_hw_init(struct rt305x_es
 	esw_w32(esw, ~RT305X_ESW_PORT_ST_CHG, RT305X_ESW_REG_IMR);
 }
 
+/* begin: add by chenzejun for wan led by 2016.6.22 */
+struct net_device *ralink_esw_netdev = NULL;
+EXPORT_SYMBOL_GPL(ralink_esw_netdev);
+/* end: add by chenzejun for wan led by 2016.6.22 */
 static irqreturn_t esw_interrupt(int irq, void *_esw)
 {
 	struct rt305x_esw *esw = (struct rt305x_esw *) _esw;
-	u32 status;
+	u32 status, carrier = 0;
 	u32 poa, i, link_status;
 	char port_num[32];
 	char *envp[] = {port_num, NULL};
@@ -754,10 +758,24 @@ static irqreturn_t esw_interrupt(int irq
                  		       kobject_uevent_env(&esw->dev->kobj, KOBJ_OFFLINE, envp);
          		              }
                                   }
+                                  if (link_status)  carrier++;
                           }
 	}
 	esw_w32(esw, status, RT305X_ESW_REG_ISR);
 
+/* begin: add by chenzejun for wan led by 2016.6.22 */
+             if (ralink_esw_netdev && (carrier > 1))
+        	{
+        		//printk("link up wan, carrier:%d\n", carrier);
+        		netif_carrier_on(ralink_esw_netdev);
+        	}
+        	else
+        	{
+        		//printk("link down wan, carrier:%d\n", carrier);
+        		netif_carrier_off(ralink_esw_netdev);
+        	}
+/* end: add by chenzejun for wan led by 2016.6.22 */
+
 	return IRQ_HANDLED;
 }
 
--- a/drivers/net/ethernet/ralink/ralink_soc_eth.c
+++ b/drivers/net/ethernet/ralink/ralink_soc_eth.c
@@ -1411,6 +1411,11 @@ static void fe_pending_work(struct work_
 	}
 }
 
+/* begin: add by chenzejun for wan led by 2016.6.22 */
+#ifdef CONFIG_NET_RALINK_ESW_RT3052
+extern struct net_device *ralink_esw_netdev;
+#endif
+/* end: add by chenzejun for wan led by 2016.6.22 */
 static int fe_probe(struct platform_device *pdev)
 {
 	struct resource *res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
@@ -1508,6 +1513,12 @@ static int fe_probe(struct platform_devi
 
 	platform_set_drvdata(pdev, netdev);
 
+/* begin: add by chenzejun for wan led by 2016.6.22 */
+#ifdef CONFIG_NET_RALINK_ESW_RT3052
+	ralink_esw_netdev = netdev;
+#endif
+/* end: add by chenzejun for wan led by 2016.6.22 */
+
 	netif_info(priv, probe, netdev, "ralink at 0x%08lx, irq %d\n",
 			netdev->base_addr, netdev->irq);
 
