--- a/drivers/net/ethernet/ralink/gsw_mt7620a.c
+++ b/drivers/net/ethernet/ralink/gsw_mt7620a.c
@@ -287,7 +287,9 @@ static irqreturn_t gsw_interrupt_mt7620(
 	struct mt7620_gsw *gsw = (struct mt7620_gsw *) priv->soc->swpriv;
 	u32 status;
 	int i, max = (gsw->port4 == PORT4_EPHY) ? (4) : (3);
-
+	char port_num[32];
+	char *envp[] = {port_num, NULL};
+	
 	status = gsw_r32(gsw, GSW_REG_ISR);
 	if (status & PORT_IRQ_ST_CHG)
 		for (i = 0; i <= max; i++) {
@@ -296,11 +298,20 @@ static irqreturn_t gsw_interrupt_mt7620(
 
 			if (link != priv->link[i]) {
 				if (link)
+				{
 					netdev_info(priv->netdev, "port %d link up (%sMbps/%s duplex)\n",
 							i, fe_speed_str((status >> 2) & 3),
 							(status & 0x2) ? "Full" : "Half");
+
+					snprintf(port_num, sizeof(port_num), "PORT_ID=%d", i);
+					kobject_uevent_env(&priv->netdev->dev.kobj, KOBJ_ONLINE, envp);
+				}
 				else
+				{
 					netdev_info(priv->netdev, "port %d link down\n", i);
+					snprintf(port_num, sizeof(port_num), "PORT_ID=%d", i);
+					kobject_uevent_env(&priv->netdev->dev.kobj, KOBJ_OFFLINE, envp);
+				}
 			}
 
 			priv->link[i] = link;
