diff -Naur a/drivers/mmc/host/mtk-mmc/board.h b/drivers/mmc/host/mtk-mmc/board.h
--- a/drivers/mmc/host/mtk-mmc/board.h	2016-10-20 14:32:51.000000000 +0800
+++ b/drivers/mmc/host/mtk-mmc/board.h	2016-10-20 12:14:11.000000000 +0800
@@ -84,7 +84,8 @@
     unsigned long  flags;            /* hardware capability flags */
     unsigned long  data_pins;        /* data pins */
     unsigned long  data_offset;      /* data address offset */
-
+	unsigned char crc_edge;
+	
     /* config gpio pull mode */
     void (*config_gpio_pin)(int type, int pull);
 
diff -Naur a/drivers/mmc/host/mtk-mmc/mt6575_sd.h b/drivers/mmc/host/mtk-mmc/mt6575_sd.h
--- a/drivers/mmc/host/mtk-mmc/mt6575_sd.h	2016-10-20 14:32:51.000000000 +0800
+++ b/drivers/mmc/host/mtk-mmc/mt6575_sd.h	2016-10-20 12:03:56.000000000 +0800
@@ -234,6 +234,7 @@
 #define MSDC_IOCON_SDR104CKS    (0x1  << 0)     /* RW */
 #define MSDC_IOCON_RSPL         (0x1  << 1)     /* RW */
 #define MSDC_IOCON_DSPL         (0x1  << 2)     /* RW */
+#define MSDC_IOCON_WDSPL        (0x1  << 8)     /* RW */
 #define MSDC_IOCON_DDLSEL       (0x1  << 3)     /* RW */
 #define MSDC_IOCON_DDR50CKD     (0x1  << 4)     /* RW */
 #define MSDC_IOCON_DSPLSEL      (0x1  << 5)     /* RW */
diff -Naur a/drivers/mmc/host/mtk-mmc/sd.c b/drivers/mmc/host/mtk-mmc/sd.c
--- a/drivers/mmc/host/mtk-mmc/sd.c	2016-10-20 14:32:51.000000000 +0800
+++ b/drivers/mmc/host/mtk-mmc/sd.c	2016-10-20 14:49:17.000000000 +0800
@@ -190,6 +190,7 @@
 	.clk_src        = 0,
 	.cmd_edge       = MSDC_SMPL_FALLING,
 	.data_edge      = MSDC_SMPL_FALLING,
+	.crc_edge      = MSDC_SMPL_FALLING,
 	.clk_drv        = 4,
 	.cmd_drv        = 4,
 	.dat_drv        = 4,
@@ -2267,7 +2268,11 @@
             INIT_MSG("SD data latch edge<%d>", hw->data_edge);            
             sdr_set_field(MSDC_IOCON, MSDC_IOCON_RSPL, hw->cmd_edge);
             sdr_set_field(MSDC_IOCON, MSDC_IOCON_DSPL, hw->data_edge);
-            //} /* for tuning debug */
+			
+            /* sdxc: set sample crc by clock falling edge . by zhangzf 20161020*/
+            if (ralink_soc != MT762X_SOC_MT7620A)
+                sdr_set_field(MSDC_IOCON, MSDC_IOCON_WDSPL, hw->crc_edge);
+        //} /* for tuning debug */
         } else { /* default value */
             sdr_write32(MSDC_IOCON,      0x00000000);
             // sdr_write32(MSDC_DAT_RDDLY0, 0x00000000);
@@ -2996,7 +3001,7 @@
 {
     int ret;
 /* +++ by chhung */
-    u32 reg;
+    u32 reg, reg1;
 
 #if defined (CONFIG_MTD_ANY_RALINK)
     extern int ra_check_flash_type(void);
@@ -3006,37 +3011,42 @@
     }
 #endif
     printk("MTK MSDC device init.\n");
+	
     mtk_sd_device.dev.platform_data = &msdc0_hw;
-if (ralink_soc == MT762X_SOC_MT7620A || ralink_soc == MT762X_SOC_MT7621AT) {
-//#if defined (CONFIG_RALINK_MT7620) || defined (CONFIG_RALINK_MT7621)
-    reg = sdr_read32((volatile u32*)(RALINK_SYSCTL_BASE + 0x60)) & ~(0x3<<18);
-//#if defined (CONFIG_RALINK_MT7620)
-	if (ralink_soc == MT762X_SOC_MT7620A)
-    		reg |= 0x1<<18;
-//#endif
-} else {
-//#elif defined (CONFIG_RALINK_MT7628)
-    /* TODO: maybe omitted when RAether already toggle AGPIO_CFG */
-    reg = sdr_read32((volatile u32*)(RALINK_SYSCTL_BASE + 0x3c));
-    reg |= 0x1e << 16;
-    sdr_write32((volatile u32*)(RALINK_SYSCTL_BASE + 0x3c), reg);
-
-    reg = sdr_read32((volatile u32*)(RALINK_SYSCTL_BASE + 0x60)) & ~(0x3<<10);
+	reg = sdr_read32((volatile u32*)(RALINK_SYSCTL_BASE + 0x60));
+	if (ralink_soc == MT762X_SOC_MT7620A || ralink_soc == MT762X_SOC_MT7621AT) {
+		reg = sdr_read32((volatile u32*)(RALINK_SYSCTL_BASE + 0x60)) & ~(0x3<<18);
+		if (ralink_soc == MT762X_SOC_MT7620A) reg |= 0x1<<18;
+		
+	/* fix MT7628 SDXC init by zhangzf 20161020 */
+	} else if (ralink_soc == MT762X_SOC_MT7628AN){
+		reg &= ~((0x3<<0)|(0x3<<6)|(0x3<<10)|(0x1<<15)|(0x3<<20)|(0x3<<24));
+		reg |=  ((0x1<<0)|(0x1<<6)|(0x1<<10)|(0x1<<15)|(0x1<<20)|(0x1<<24));
 #if defined (CONFIG_MTK_MMC_EMMC_8BIT)
-    reg |= 0x3<<26 | 0x3<<28 | 0x3<<30;
-    msdc0_hw.data_pins      = 8,
+		reg |= 0x3<<26 | 0x3<<28 | 0x3<<30;
+		msdc0_hw.data_pins	= 8;
 #endif
-//#endif
-}
+		reg1 = sdr_read32((volatile u32*)(RALINK_SYSCTL_BASE + 0x1340));
+		reg1 |= (0x1<<11); //Normal mode(AP mode), SDXC CLK=PAD_GPIO0=GPIO11, driving = 8mA
+		sdr_write32((volatile u32*)(RALINK_SYSCTL_BASE + 0x1340), reg1);
+	} else {
+		/* TODO: maybe omitted when RAether already toggle AGPIO_CFG */
+		reg = sdr_read32((volatile u32*)(RALINK_SYSCTL_BASE + 0x3c));
+		reg |= 0x1e << 16;
+		sdr_write32((volatile u32*)(RALINK_SYSCTL_BASE + 0x3c), reg);
+
+		reg = sdr_read32((volatile u32*)(RALINK_SYSCTL_BASE + 0x60)) & ~(0x3<<10);
+	}
+	
     sdr_write32((volatile u32*)(RALINK_SYSCTL_BASE + 0x60), reg);
-    //platform_device_register(&mtk_sd_device);
-/* end of +++ */
 
     ret = platform_driver_register(&mt_msdc_driver);
+	
     if (ret) {
         printk(KERN_ERR DRV_NAME ": Can't register driver");
         return ret;
     }
+	
     printk(KERN_INFO DRV_NAME ": MediaTek MT6575 MSDC Driver\n");
 
 #if defined (MT6575_SD_DEBUG)
